### ISSUE TITLE
Contrastive Input Pipeline

### ISSUE THREAD

**Issue by philippe-eecs** [google-research/big_vision#75]
Hi, this is an amazing codebase for big-vision tasks!

I wanted to re-implement MOCO, but I was unsure how to modify the input pipeline + data augmentation to allow for applying independent random augmentations to the same image. Is there a simple way to do this? 

My current implementation just applies the augmentations after the batch of images leaves the input pipeline (without any augmentation), but this requires me to write new data augmentation functions in Jax, which isn't ideal. Do you have any ideas on how to integrate this into the input pipeline?

Any help would be appreciated and let me know if you need more information.

**Comment by akolesnikoff** [comment]
I think the easiest solution is to hack input_pipeline a bit and duplicate images in the input pipeline. `flat_map` operation can be used for this, see the relevant question on stack overflow: https://stackoverflow.com/questions/61754089/how-to-repeat-a-tf-data-dataset

**Comment by philippe-eecs** [comment]
`def duplicate_element(x):
    # Create a dataset that contains two copies of x
    return tf.data.Dataset.from_tensors(x).repeat(2)

def make_for_train(
    data, preprocess_fn, batch_size,
    shuffle_buffer_size, cache_raw=False,
    num_parallel_calls=100, prefetch=2,
    duplicate_images=False,
    *,
    pre_filter_fn=None, post_filter_fn=None,
    pack=None,
):
  """Makes an input pipeline for training."""
  # Use data filtering at your own risk: the actual split sizes won't be known
  # in advance, so epoch-based things won't work correctly.

  data = _add_tpu_host_options(data)

  data = data.filter(pre_filter_fn) if pre_filter_fn else data
  data = data.cache() if cache_raw else data
  data = data.repeat(None)  # repeat data indefinitely
  data = data.shuffle(shuffle_buffer_size) if shuffle_buffer_size else data

  if duplicate_images:
    #from https://stackoverflow.com/questions/61754089/how-to-repeat-a-tf-data-dataset
    #Should give positive pairs
    data = data.flat_map(duplicate_element)

  data = data.map(preprocess_fn, num_parallel_calls=num_parallel_calls)
  data = data.filter(post_filter_fn) if post_filter_fn else data

  data = sequence_packing.pack_dataset(data, pack) if pack else data

  # Drop remainder makes shape fully static, so we can later use it if needed.
  if batch_size:
    data = data.batch(batch_size // jax.process_count(), drop_remainder=True)
  if prefetch:  # None means autotune, but we never want that.
    data = data.prefetch(prefetch)
  return data`
  
  
This worked. Thanks for the idea!


Test on cifar10: first 64 "samples"


![image](https://github.com/google-research/big_vision/assets/77702076/9b868db5-bfc8-49d5-89fa-da07db18dadc)
